### PR TITLE
Implement caching for license texts

### PR DESCRIPTION
**Pull request by alois31**
License texts are not expected to ever change. Cache them so that
network access is only required when first dowloading a new license.

Signed-off-by: Alois Wohlschlager <alois1@gmx-topmail.de>

### CODE REVIEW

File: src/reuse/download.py [hunk 1]
```diff
@@ -38,13 +39,21 @@ def download_license(spdx_identifier: str) -> str:
...
+    cachehome = os.environ["XDG_CACHE_HOME"] if "XDG_CACHE_HOME" in os.environ else Path(os.environ["HOME"]) / ".cache"
+    cachedir = Path(cachehome) / "reuse"
+    try:
+        with open(cachedir / filename, "r") as cachefile:
+            return cachefile.read()
+    except FileNotFoundError:
+        pass
     response = requests.get(url)
     if response.status_code == 200:
+        os.makedirs(cachedir, exist_ok=True)
```
Comment (siiptuo) [inline src/reuse/download.py:54]:
```suggestion
        cachedir.mkdir(parents=True, exist_ok=True)
```

**Review by siiptuo — CHANGES_REQUESTED** [review]
The license text are mostly static but there has been some changes to formatting over the years. However, caching the licenses is a good idea and here's some suggestion mostly on code style.

File: src/reuse/download.py [hunk 2]
```diff
@@ -38,13 +39,21 @@ def download_license(spdx_identifier: str) -> str:
...
     :return: The license text.
     """
+    filename = "".join((spdx_identifier, ".txt"))
     # This is fairly naive, but I can't see anything wrong with it.
-    url = urljoin(
-        _SPDX_REPOSITORY_BASE_URL, "".join((spdx_identifier, ".txt"))
-    )
-    # TODO: Cache result?
+    url = urljoin(_SPDX_REPOSITORY_BASE_URL, filename)
+    cachehome = os.environ["XDG_CACHE_HOME"] if "XDG_CACHE_HOME" in os.environ else Path(os.environ["HOME"]) / ".cache"
```
[the remark this answers is not in the document]
Comment (mxmehl) [reply src/reuse/download.py:45]: Would that also work on Windows and MacOS machines? We have a few users on other operating systems.

**Comment by mxmehl** [comment]
As @siiptuo mentioned, license texts in SPDX tend to change here and there, often fixing bugs that users benefit from. I would be fine with some time-limited caching, e.g. a few days or bound to the REUSE tool version, so that with a new version the cached license texts are refreshed.

However, we should cover other operating systems. Is `XDG_CACHE_HOME` understood on MacOS and Windows as well?

File: src/reuse/download.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (siiptuo) [reply src/reuse/download.py:45]: I think that with `Path.home` it should work on all platforms but this should be tested. More ideal solution would be to use a library like [appdirs](https://pypi.org/project/appdirs/).

**Comment by carmenbianca** [comment]
> However, we should cover other operating systems. Is `XDG_CACHE_HOME` understood on MacOS and Windows as well?

[appdirs](https://pypi.org/project/appdirs/) covers this use case. It's a fairly common dependency, but I'm not sure we should add it.

File: src/reuse/download.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mxmehl) [reply src/reuse/download.py:45]: The CI has a number of other issues currently, caused by an update of an unpinned dependency of Jinja, markupsafe. Was it this error?

We are usually very cautious of adding new dependencies. However, if it's really necessary, their should at least be packaged in Debian and Fedora.

[the remark this answers is not in the document]
Comment (alois31) [reply src/reuse/download.py:45]: It was actually about appdirs itself. It is packaged (at least) on all of Debian, Fedora and openSUSE.

**Comment by seabass-labrax** [comment]
How about a cache based on the current SPDX License List version, rather than a duration? This could be done by scraping the version label on https://spdx.org/licenses, fetching the latest tag from the SPDX License List Git repository (although that wouldn't work if the Git command isn't present) or by making a GitHub API request to find the latest release.

**Comment by floriansnow** [comment]
So how should we proceed with this?  I think it's nice to heave, but probably not a deal breaker for most use cases.  There are still a few open questions here and by now also merge conflicts.  Should we invest the time or table this for now because it's probably "just" nice to have?

**Comment by mxmehl** [comment]
I would also consider it nice-to-have, but as we learnt in this thread, some questions/issues would have to be resolved first:

- Logic to detect the cache dir. `appdirs` works but saw its last release in 2020. Is there a successor that's also widely available in distros?
- Limited caching period, based on SPDX License List version (or alternatively REUSE tool version). Both are not perfect indicators, but at least the former should be quite sufficient.

**Comment by linozen** [comment]
IMHO, the complexity involved in implementing this properly (making sure licenses are up-to-date etc.) while adding only well-maintained and broadly supported dependencies stands in no relationship to the benefit this could generate for users. If the only use-case here remains the one outlined by @alois31, I am in favor of closing this.

**Pull request closed without merging**